### PR TITLE
[CINN] Dump more compilation result and optimize parallel compiler flags

### PR DESCRIPTION
**Pull request by FisherWY**
### PR types
Others

### PR changes
Others

### Description
Pcard-72511

1. `Parallel Compiler`：
    - 合并`FLAGS_cinn_parallel_compile_size`和`FLAGS_cinn_parallel_compile_thread`，通过`FLAGS_cinn_parallel_compile_thread`即可指定编译时使用的线程数，所有的`fusion_groups`将会平均分配到可用的线程上
    - 增强编译完成后返回的信息，除`instruction`外，将`lowered_function`、`source_code`、`source_ptx`返回，供上层进一步使用
2. Debug信息：
    - 新增`FLAGS_ cinn_dump_group_lowered_func`、`FLAGS_cinn_dump_group_source_code`、`FLAGS_ cinn_dump_group_ptx`、`FLAGS_ cinn_dump_group_instruction`，可分别按`fusion_groups`储存编译的每个阶段中的中间代码
    - 重新整理`graph_visualization`，所有的可视化图、单测代码均能正确分组储存
3. Bug修复：
    - 修复`MakeDirectory`不能正确创建文件夹的问题
4. 其他：
    - 清除了一些无用代码

### CODE REVIEW

**Review by BiynXu — APPROVED** [review]
LGTM

File: paddle/cinn/backends/compiler.cc [hunk 1]
```diff
@@ -29,13 +30,125 @@
...
 namespace backends {
 using ir::Module;
 
 static constexpr int DebugLogMaxLen = 30000;
 
+void DumpCompilationInfo::DumpLoweredFunc() {
+  if (FLAGS_cinn_dump_group_lowered_func.empty()) {
+    return;
+  }
+  for (int idx = 0; idx < info_.lowered_funcs.size(); idx++) {
```
Comment (zhhsplendid) [inline paddle/cinn/backends/compiler.cc]: Prefer `++idx`

Reference: https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement

File: paddle/cinn/backends/compiler.cc [hunk 2]
```diff
@@ -29,13 +30,125 @@
...
+          "%s/%s", dump_path.c_str(), "lowered_function.txt");
+      VLOG(7) << "Dump lower functions to: " << dump_file;
+      std::ofstream of(dump_file, std::ios_base::out);
+      if (of.is_open()) {
+        of << info_.lowered_funcs[idx].front();
+      } else {
+        LOG(WARNING) << "Failed to open file: " << dump_file
+                     << ", please check your path.";
+      }
+      of.close();
```
Comment (zhhsplendid) [inline paddle/cinn/backends/compiler.cc]: Based on the C++ reference, close on an `ofstream` which is not opened may throw [ios_base::failure](https://cplusplus.com/ios_base::failure)

Could you put `of.close()` in `if (of.is_open())` condition branch to make code safer? Only close after you open it. : -)

Reference: https://cplusplus.com/reference/fstream/ofstream/close/

File: paddle/cinn/backends/compiler.cc [hunk 3]
```diff
@@ -29,13 +30,125 @@
...
+          utils::StringFormat("%s/%s", dump_path.c_str(), "source_code.cu");
+      VLOG(7) << "Dump source code to: " << dump_file;
+      std::ofstream of(dump_file, std::ios_base::out);
+      if (of.is_open()) {
+        of << info_.source_codes[idx];
+      } else {
+        LOG(WARNING) << "Failed to open file: " << dump_file
+                     << ", please check your path.";
+      }
+      of.close();
```
Comment (zhhsplendid) [inline paddle/cinn/backends/compiler.cc]: Is this function has lots of common code with last function? If so, can we extract some common codes as util function?

File: paddle/cinn/backends/compiler.h [hunk 4]
```diff
@@ -32,6 +33,34 @@
...
 
+/**
+ * A class for dumping the code after compilation.
+ * Use FLAGS_cinn_dump_group_lowered_func to specify the directory to dump
+ * lowered function. Use FLAGS_cinn_dump_group_source_code to specify the
+ * directory to dump the source code. Use FLAGS_cinn_dump_group_ptx to specify
+ * the directory to dump ptx. Use FLAGS_cinn_dump_group_instruction to specify
+ * the directory to dump instruction.
+ */
+class DumpCompilationInfo {
```
Comment (zhhsplendid) [inline paddle/cinn/backends/compiler.h]: Although there are exceptions, but modern software engineer encourages class to be "noun" while method to be "verb" (so we call `object` as the instance of class, not `action`)

For example:
`Executor.run`
`Logger.log`

So I suggest you change class name to `CompilationInfoDumper`

File: paddle/cinn/hlir/framework/graph.cc [hunk 5]
```diff
@@ -308,66 +308,40 @@ void Graph::VisualizeGroupedGraph(
...
-
-  Summary(groups, viz_path_);
-  WriteToFile(viz_path_ + "grouped_graph.dot",
-              VisualizeGraph(groups, fetch_var_ids));
-
-  {
-    // save each group's graphviz dot file
-    std::string group_path = viz_path_ + "/groups/";
+  const auto& group_dots = VisualizeGroups(groups, fetch_var_ids);
+  for (int idx = 0; idx < groups.size(); idx++) {
```
Comment (zhhsplendid) [inline paddle/cinn/hlir/framework/graph.cc]: Same as above, prefer `++idx`

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Pull request merged**